In [2]:
import numpy as np
import pandas as pd
from pydataset import data
np.random.seed(117)

## 1. Copy the code from the lesson to create a dataframe full of student grades.

In [3]:
students = ['Sally',
            'Jane',
            'Suzie',
            'Billy',
            'Ada',
            'John',
            'Thomas',
            'Marie',
            'Albert',
            'Richard',
            'Isaac',
            'Alan']

In [4]:
pd.Series(students)

0       Sally
1        Jane
2       Suzie
3       Billy
4         Ada
5        John
6      Thomas
7       Marie
8      Albert
9     Richard
10      Isaac
11       Alan
dtype: object

In [5]:
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

In [6]:
df_dict = {'name':students,
           'math': math_grades,
           'english': english_grades,
           'reading':reading_grades}

In [7]:
df = pd.DataFrame(df_dict)

In [8]:
df

name  math  english  reading
0     Sally    76       71       79
1      Jane    77       98       97
2     Suzie    83       62       84
3     Billy    79       82       65
4       Ada    93       82       64
5      John    97       93       80
6    Thomas    77       74       94
7     Marie    71       76       91
8    Albert    94       91       74
9   Richard    87       72       91
10    Isaac    62       67       68
11     Alan    86       97       85

### a. Create a column named passing_english that indicates whether each student has a passing grade in english.

In [9]:
df['passing_english'] = df.english >= 70

In [10]:
df

name  math  english  reading  passing_english
0     Sally    76       71       79             True
1      Jane    77       98       97             True
2     Suzie    83       62       84            False
3     Billy    79       82       65             True
4       Ada    93       82       64             True
5      John    97       93       80             True
6    Thomas    77       74       94             True
7     Marie    71       76       91             True
8    Albert    94       91       74             True
9   Richard    87       72       91             True
10    Isaac    62       67       68            False
11     Alan    86       97       85             True

### b. Sort the english grades by the passing_english column. How are duplicates handled?

In [11]:
df.sort_values(by='passing_english')

name  math  english  reading  passing_english
2     Suzie    83       62       84            False
10    Isaac    62       67       68            False
0     Sally    76       71       79             True
1      Jane    77       98       97             True
3     Billy    79       82       65             True
4       Ada    93       82       64             True
5      John    97       93       80             True
6    Thomas    77       74       94             True
7     Marie    71       76       91             True
8    Albert    94       91       74             True
9   Richard    87       72       91             True
11     Alan    86       97       85             True

### c. Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)

In [12]:
df.sort_values(by='name').sort_values(by='passing_english')

name  math  english  reading  passing_english
10    Isaac    62       67       68            False
2     Suzie    83       62       84            False
4       Ada    93       82       64             True
11     Alan    86       97       85             True
8    Albert    94       91       74             True
3     Billy    79       82       65             True
1      Jane    77       98       97             True
5      John    97       93       80             True
7     Marie    71       76       91             True
9   Richard    87       72       91             True
0     Sally    76       71       79             True
6    Thomas    77       74       94             True

### d. Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [13]:
df.sort_values(by='passing_english').sort_values(by='english')

name  math  english  reading  passing_english
2     Suzie    83       62       84            False
10    Isaac    62       67       68            False
0     Sally    76       71       79             True
9   Richard    87       72       91             True
6    Thomas    77       74       94             True
7     Marie    71       76       91             True
3     Billy    79       82       65             True
4       Ada    93       82       64             True
8    Albert    94       91       74             True
5      John    97       93       80             True
11     Alan    86       97       85             True
1      Jane    77       98       97             True

### e. Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [14]:
df['overall grade'] = (df.english + df.math + df.reading) / 3

In [15]:
df

name  math  english  reading  passing_english  overall grade
0     Sally    76       71       79             True      75.333333
1      Jane    77       98       97             True      90.666667
2     Suzie    83       62       84            False      76.333333
3     Billy    79       82       65             True      75.333333
4       Ada    93       82       64             True      79.666667
5      John    97       93       80             True      90.000000
6    Thomas    77       74       94             True      81.666667
7     Marie    71       76       91             True      79.333333
8    Albert    94       91       74             True      86.333333
9   Richard    87       72       91             True      83.333333
10    Isaac    62       67       68            False      65.666667
11     Alan    86       97       85             True      89.333333

## 2. Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [16]:
mpg = data('mpg')

In [17]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

### a. How many rows and columns are there?

In [18]:
mpg.shape

(234, 11)

### b. What are the data types of each column?

In [19]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

### c. Summarize the dataframe with .info and .describe

In [20]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [21]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

### d. Rename the cty column to city.

In [22]:
mpg = mpg.rename(columns={'cty': 'city'})

### e. Rename the hwy column to highway.

In [23]:
mpg = mpg.rename(columns={'hwy': 'highway'})

### f. Do any cars have better city mileage than highway mileage?

In [24]:
better_city_milage = (mpg.city > mpg.highway).sum()

In [25]:
better_city_milage

0

No

### g. Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

In [26]:
mpg['mileage_difference'] = (mpg.highway - mpg.city)

In [27]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

### h. Which car (or cars) has the highest mileage difference?

In [28]:
mpg.sort_values(by='mileage_difference', ascending = False).head(2)

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   

    fl       class  mileage_difference  
107  c  subcompact                  12  
223  d  subcompact                  12

Both the 2008 Honda Civic and the 1999 Volkswagen New Beetle have the highest mileage differance at 12 miles.

### i. Which compact class car has the lowest highway mileage? The best?

In [29]:
mpg.sort_values('highway', ascending=False)

manufacturer                model  displ  year  cyl       trans drv  city  \
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f    35   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f    33   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f    29   
197       toyota              corolla    1.8  2008    4  manual(m5)   f    28   
106        honda                civic    1.8  2008    4    auto(l5)   f    25   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   

     highway fl       class  mileage_difference  
222       44  d  subcompact                   9  
213       44  d     compact                  11  
223       41  d  subcompact                  12  
197       37  r     compact                   9  
106       36  r  subcompact                  11  
..       ... ..         ...                 ...  
66        12  e      pickup                   3  
60        12  e         suv                   3  
127       12  e         suv                   3  
70        12  e      pickup                   3  
55        12  e      pickup                   3  

[234 rows x 12 columns]

In [30]:
max_hwy_compact = mpg[mpg['class'] == 'compact'].describe()['highway']['max']

In [31]:
min_hwy_compact = mpg[mpg['class'] == 'compact'].describe()['highway']['min']

In [32]:
mpg[(mpg.highway == max_hwy_compact) | (mpg.highway == min_hwy_compact)]

manufacturer         model  displ  year  cyl       trans drv  city  \
18          audi    a6 quattro    4.2  2008    8    auto(s6)   4    16   
25     chevrolet      corvette    5.7  1999    8    auto(l4)   r    15   
47         dodge   caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48         dodge   caravan 2wd    4.0  2008    6    auto(l6)   f    16   
97          ford       mustang    4.6  2008    8  manual(m5)   r    15   
165       subaru  forester awd    2.5  2008    4    auto(l4)   4    18   
213   volkswagen         jetta    1.9  1999    4  manual(m5)   f    33   
220   volkswagen         jetta    2.8  1999    6    auto(l4)   f    16   
222   volkswagen    new beetle    1.9  1999    4  manual(m5)   f    35   

     highway fl       class  mileage_difference  
18        23  p     midsize                   7  
25        23  p     2seater                   8  
47        23  r     minivan                   7  
48        23  r     minivan                   7  
97        23  r  subcompact                   8  
165       23  p         suv                   5  
213       44  d     compact                  11  
220       23  r     compact                   7  
222       44  d  subcompact                   9

1999 Volkswagen Jetta with manual transmission and 4 cylinders has the best highway milage at 44. Oddly enough, the 1999 Volkswagen Jetta but with auto transmission and 6 cylinders also had the worst(lowest) highway milage at 23.

### j. Create a column named average_mileage that is the mean of the city and highway mileage.

In [33]:
mpg['average_mileage'] = ((mpg.highway + mpg.city) / 2)

In [34]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

### k. Which dodge car has the best average mileage? The worst?

In [35]:
max_avg_mileage_dodge = mpg[mpg['manufacturer'] == 'dodge'].describe()['average_mileage']['max']

In [36]:
min_avg_mileage_dodge = mpg[mpg['manufacturer'] == 'dodge'].describe()['average_mileage']['min']

In [37]:
mpg[(mpg.average_mileage == max_avg_mileage_dodge) | (mpg.average_mileage == min_avg_mileage_dodge)]

manufacturer                model  displ  year  cyl       trans drv  city  \
5           audi                   a4    2.8  1999    6    auto(l5)   f    16   
13          audi           a4 quattro    2.8  1999    6  manual(m5)   4    17   
14          audi           a4 quattro    3.1  2008    6    auto(s6)   4    17   
17          audi           a6 quattro    3.1  2008    6    auto(s6)   4    17   
24     chevrolet             corvette    5.7  1999    8  manual(m6)   r    16   
26     chevrolet             corvette    6.2  2008    8  manual(m6)   r    16   
38         dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   
156      pontiac           grand prix    3.8  1999    6    auto(l4)   f    16   
161       subaru         forester awd    2.5  1999    4    auto(l4)   4    18   
232   volkswagen               passat    2.8  1999    6    auto(l5)   f    16   

     highway fl    class  mileage_difference  average_mileage  
5         26  p  compact                  10             21.0  
13        25  p  compact                   8             21.0  
14        25  p  compact                   8             21.0  
17        25  p  midsize                   8             21.0  
24        26  p  2seater                  10             21.0  
26        26  p  2seater                  10             21.0  
38        24  r  minivan                   6             21.0  
55        12  e   pickup                   3             10.5  
60        12  e      suv                   3             10.5  
66        12  e   pickup                   3             10.5  
70        12  e   pickup                   3             10.5  
127       12  e      suv                   3             10.5  
156       26  p  midsize                  10             21.0  
161       24  r      suv                   6             21.0  
232       26  p  midsize                  10             21.0

The 1999 Dodge Caravan 2WD with 4 cyclinders and auto transmission has the highest average mileage at 21. The 2008 Dodge Ram 1500 pickup 4wd has the worst average milage at 10.5.

## 3. Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

In [38]:
data('Mammals')

weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    130.000   70.0    False     False
25    120.000   80.0    False     False
26    120.000   61.0    False     False
27    110.000   33.0    False     False
28    100.000   64.0    False     False
29     85.000   55.0    False     False
30     80.000   65.0    False     False
31     72.000   56.0    False     False
32     70.000   45.0    False     False
33     65.000   60.0    False     False
34     62.000   81.0    False     False
35     50.000  100.0    False     False
36     50.000   60.0    False     False
37     50.000   40.0    False     False
38     50.000   47.0    False     False
39     37.000  105.0    False     False
40     35.000   80.0    False     False
41     34.000   97.0    False     False
42     30.000   97.0    False     False
43     30.000   80.0    False     False
44     30.000   45.0    False     False
45     20.000   81.0    False     False
46    400.000   40.0    False     False
47    300.000   48.0    False     False
48    230.000   56.0    False     False
49    150.000   59.0    False     False
50    135.000   48.0    False     False
51     65.000   65.0    False     False
52     60.000   60.0    False     False
53     55.000  110.0    False     False
54     45.000   50.0    False     False
55     40.000   64.0    False     False
56     25.000   67.0    False     False
57     20.000   70.0    False     False
58     16.000   65.0    False     False
59     12.000   24.0    False      True
60     11.000   30.0    False      True
61     10.000   56.0    False     False
62      7.000   60.0    False     False
63      6.000   72.0    False     False
64      5.000   64.0    False     False
65      5.000   27.0    False      True
66      3.000   16.0    False      True
67    127.000   32.0    False     False
68     70.000   40.0    False      True
69     13.000   37.0    False      True
70      9.000    3.2    False      True
71      4.000   16.0    False     False
72      0.600   36.0    False     False
73      0.600   20.0    False     False
74      0.550   27.0    False     False
75      0.500   18.0    False     False
76      0.400   20.0    False     False
77      0.300   13.0    False     False
78      0.250    9.7    False     False
79      0.220   15.0    False     False
80      0.110    9.0    False     False
81      0.100   17.0    False     False
82      0.056   21.0     True     False
83      0.050   11.0    False     False
84      0.045   16.0    False     False
85      0.035   32.0     True     False
86      0.035   14.0     True     False
87      0.030    6.8    False     False
88      0.030    9.1    False     False
89      0.025   11.0    False     False
90      0.025    8.6    False     False
91      0.018    8.9    False     False
92      0.016   13.0    False     False
93      0.100    4.0    False     False
94      0.100    2.4    False     False
95      0.016    3.6    False     False
96      4.600   64.0     True     False
97      4.400   72.0     True     False
98      4.000   72.0     True     False
99      3.500   56.0     True     False
100    

In [39]:
mammals = data('Mammals')

### a. How many rows and columns are there?

In [40]:
mammals.shape

(107, 4)

107 rows, 4 columns

### b. What are the data types?

In [41]:
mammals.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

### c. Summarize the dataframe with .info and .describe

In [42]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [43]:
mammals.describe()

weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

### d. What is the the weight of the fastest animal?

In [44]:
mammals.sort_values(by='speed',ascending = False).head(1)

weight  speed  hoppers  specials
53    55.0  110.0    False     False

55 kg

### e. What is the overal percentage of specials?

In [46]:
spec_mamm = mammals['specials'].value_counts()[True]

In [47]:
spec_mamm

10

In [48]:
tot_mamm = (mammals['specials'].value_counts()[True]) + mammals['specials'].value_counts()[False]

In [49]:
per_spec = spec_mamm / tot_mamm

In [62]:
#this is done
print(f'The overall percentage of specials is {round(per_spec,4) * 100}%')

The overall percentage of specials is 9.35%


In [51]:
# ignore this
#just trying to do eveything in one line
#print(f' The overall percentage of specials is {round(mammals['specials'].value_counts()[True] / (mammals['specials'].value_counts()[True]) + mammals['specials'].value_counts()[False],3) * 100}%')

### f. How many animals are hoppers that are above the median speed? What percentage is this?

7 animals, 6.54%

In [52]:
mammals['speed'].describe()

count    107.000000
mean      46.208411
std       26.716778
min        1.600000
25%       22.500000
50%       48.000000
75%       65.000000
max      110.000000
Name: speed, dtype: float64

In [53]:
med_speed = mammals['speed'].describe()['50%']

In [54]:
med_speed

48.0

In [55]:
hopp_mamm = mammals['hoppers'].value_counts()[True]

In [56]:
mammals[(mammals['speed'] > med_speed) & (mammals['hoppers'])]

weight  speed  hoppers  specials
96      4.6   64.0     True     False
97      4.4   72.0     True     False
98      4.0   72.0     True     False
99      3.5   56.0     True     False
100     2.0   64.0     True     False
101     1.9   56.0     True     False
102     1.5   50.0     True     False

In [57]:
num_hop_above_med = mammals[(mammals['speed'] > med_speed) & (mammals['hoppers'])].count()['hoppers']

In [58]:
num_hop_above_med

7

In [59]:
percent_of_hop_above_med = num_hop_above_med / tot_mamm

In [60]:
print(f'The overall percentage of hoppers with a speed above the median is {round(percent_of_hop_above_med,4) * 100}%')

The overall percentage of hoppers with a speed above the median is 6.54%
